### PR TITLE
kqueue: ignore ENOENT errors when EV_EOF

### PR DESCRIPTION
**Pull request by indutny**
Some sockets are automatically removed from kqueue upon their
termination, and doing EV_DELETE results in ENOENT. It could be safely
ignored as it isn't a big deal anyway.

This blocks me from using tlsnappy with node 0.9.x, as it `abort()`s suddenly on a benchmark.

### CODE REVIEW

File: src/unix/kqueue.c [hunk 1]
```diff
@@ -197,7 +197,10 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
           /* TODO batch up */
           struct kevent events[1];
           EV_SET(events + 0, fd, ev->filter, EV_DELETE, 0, 0, 0);
-          if (kevent(loop->backend_fd, events, 1, NULL, 0, NULL)) abort();
+          if (kevent(loop->backend_fd, events, 1, NULL, 0, NULL) &&
+              !(ev->flags & EV_EOF)) {
```
Comment (bnoordhuis) [inline src/unix/kqueue.c]: Shouldn't you check for ENOENT here?

Comment (indutny) [inline src/unix/kqueue.c]: Yes, I can do this too...

Comment (indutny) [inline src/unix/kqueue.c]: Force pushed.

**Comment by indutny** [comment]
Ok, I nailed it down... Please refer commit message. /cc @bnoordhuis

File: src/unix/kqueue.c [hunk 2]
```diff
@@ -167,6 +167,15 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
       fd = ev->ident;
       w = loop->watchers[fd];
 
+      /* File descriptor might be closed during callback,
+       * all events that was reported before callback are not valid in such
+       * cases and can be safely ignored
+       */
+      if (w != ev->udata) {
```
Comment (bnoordhuis) [inline src/unix/kqueue.c]: What happens if the fd is closed, then reopened and added with the same watcher?

Also, this effectively turns the NULL check below into dead code. Probably not what you want.

Comment (indutny) [inline src/unix/kqueue.c]: Actually, after sleeping well, I think that ignoring ENOENT is enough.

**Comment by indutny** [comment]
Force pushed. @bnoordhuis

**Pull request merged**